File: python/Card.py
"""Representation of a standard playing card.
"""

import CardSuit
import CardValue

class Card():
    """Representation of a standard playing card.

    Attributes:
        suit: CardSuit
        value: CardValue
    """

    def __init__(self, suit: CardSuit, value: CardValue):
        """Initialize the playing card with the given values."""
        self.suit = suit
        self.value = value
    

    def __str__(self):
        return (str(self.suit) + str(self.value))

    
    def __eq__(self, other):
        return ((self.suit == other.suit) and (self.value == other.value))
    
    def __ne__(self, other):
        return ((self.suit != other.suit) or (self.value != other.value))
    
    def __lt__(self, other):
        if(self.suit == other.suit):
            return self.value < other.value
        else:
            return self.suit < other.suit
    
    def __le__(self, other):
        if(self.suit == other.suit):
            return self.value <= other.value
        else:
            return self.suit <= other.suit
    
    def __gt__(self, other):
        if(self.suit == other.suit):
            return self.value > other.value
        else:
            return self.suit > other.suit
    
    def __ge__(self, other):
        if(self.suit == other.suit):
            return self.value >= other.value
        else:
            return self.suit >= other.suit

__author__ = "Garrett Kamila Crayton (garrett.crayton@gmail.com)"
__date__ = "2021 April 11"
__credits__ = "Garrett Kamila Crayton (garrett.crayton@gmail.com)"
__version__ = "2.0"

File: java/src/Card.java
import java.util.Comparator;

/**
 * Represents a playing card
 */
public class Card
{
    private static boolean initRan;
    protected static CardComparator comparator;
    protected Suit suit;
    protected Rank rank;

    /**
     * The process for comparing two <code>Cards</code> (used for sorting).
     *
     * @author Garrett Kamila Crayton
     * @version 1.0
     */
    private static class CardComparator implements Comparator<Card>
    {
        /**
         * Compare the two given cards.
         *
         * @param card1 the card being compared against
         * @param card2 the card being compared to
         * @return      a negative integer, zero, or positive integer depending on whether card1 is less
         *              than, equal to, or greater than card2
         */
        public int compare(Card card1, Card card2)
        {
            int compare;
            if(card1.suit == card2.suit)
            {
                compare = card1.rank.compareTo(card2.rank);
            }
            else
            {
                compare = card1.suit.compareTo(card2.suit);
            }
            return compare;
        }
    }

    /**
     * Constructs a new <code>Card</code> object
     *
     * @param suit  the <code>Suit</code> of the <code>Card</code>
     * @param rank  the <code>Rank</code> of the <code>Card</code>
     */
    public Card(Suit suit, Rank rank)
    {
        this.suit = suit;
        this.rank = rank;

        init();
    }

    private static void init()
    {
        if(!initRan)
        {
            initRan = true;
            comparator = new CardComparator();
        }
    }

    /**
     * Return the string representation of this <code>Card</code>.
     * <p>
     * This method simply returns the combination of the string representation of the
     * card's <code>Suit</code> and the string representation of the card's <code>Rank</code>.
     *
     * @return  the string representation of the <code>Card</code>
     */
    public String toString()
    {
        return (suit.toString() + rank.toString());
    }
}
